level-3 ready

Pancfly opened the pull request.

Pancfly committed:
level-3 ready

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.
    **Комментарий: Вы правильно делаете, что предоставляете пользователю возможность видеть результат своего последнего поиска (с расстановкой актуального состояния лайков) при повторном входе на страницу фильмов. Но, статус чекбокса «Короткометражки» высвечивается неправильный при повторном входе  (см. снимок). Надо сохранять в localstogage и выводить в форме поиска и строку запроса на поиск фильмов, соответствующую этому последнему поиску, и фильтрацию по короткометражкам, если она была, и правильный статус самого чекбокса «Короткометражки».**
![Снимок экрана в 2022-02-10 12-36-25](https://user-images.githubusercontent.com/19558891/153363429-190dfdf9-a573-4a92-a790-b00e5d2df44c.png)
     **Комментарий: Если были найдены короткометражки, статус чекбокса «Короткометражки» после повторного входа на страницу фильмов начинает не соответствовать результату поиска при следующих фильтрах и поисках (см. снимок).**
![Снимок экрана в 2022-02-10 12-37-43](https://user-images.githubusercontent.com/19558891/153363502-08a5c9d0-516d-468d-9c38-8597e82f4d5f.png) 
     - Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
	**Комментарий: При вводе в поле email адреса почты без точки перед именем домена верхнего уровня можно выполнить запрос на сервер. Однако, сервер возвращает статус ошибки и входа не происходит. Нужно, чтобы валидация почты и пароля и на серверной и на клиентской части приложения, была одинаковой, и при невалидных данных запрос вообще не посылался, а пользователю выдавалось сообщение об ошибке.**
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.
	**Комментарий: При редактировании данных профиля можно ввести текущие данные и сохранить неизменённые данные много раз, а нужно сделать так, чтобы можно было сохранить только изменённые данные по сравнению с текущими данными пользователя, которые содержатся в данный момент в стейте и контексте.**  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
  **Комментарий: При  повторном входе на страницу «Сохранённые фильмы» сохранённые карточки исчезают (см. снимок). Появляются снова только после перезагрузки страницы.**
![Снимок экрана в 2022-02-10 12-29-40](https://user-images.githubusercontent.com/19558891/153363623-7fd9ddb2-e599-4d0f-8b25-a4103c586b3a.png)

  **Комментарий: Для нового пользователя отображаются активные лайки и карточки, которые сохранял предыдущий пользователь. Каждый пользователь должен иметь возможность видеть только свои карточки и работать только со своими карточками.**
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [ ] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [ ] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [ ] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: длительность короткометражек, количество фильмов, отображаемое при нажатии кнопки "Еще", другие постоянные данные нужно поместить в константы с понятными названиями и вынести в файл констант, чтобы в коде не было "магических" значений.**
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 49.

Pancfly committed:
fix: after code review

constants
pattern e`mail
FilterCheckbox status
Profile updated
Navigation to Movies&SavedMovies

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными. 
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [ ] <!--4--> пользователь получает сообщение в случае любой ошибки;
  **Комментарий: если не возможно сохранить карточку, запросы на авторизацию или регистрацию завершились ошибкой, пользователь должен увидеть сообщение.** 
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 94.

## У Вас отличный дипломный проект!

Чтобы красиво оформить файл README.md, где рассказать о собственной работе над проектом и собственных достижениях, можно воспользоваться инструментом Markdown. Описание языка Markdown есть в Ваших лекциях по этой ссылке:
https://praktikum.yandex.ru/learn/web/courses/dbf98e55-0f76-444b-850c-4538708ad571/sprints/1425/topics/b4072eed-2089-45c0-9382-98ea71202341/lessons/939aeab7-0508-49e8-bcf2-199d4dbf74f7/.

Ссылки на интересные материалы и подкасты:
[habr](https://habr.com/ru/)
https://soundcloud.com/everyonecan
https://soundcloud.com/podlodka
https://soundcloud.com/devschacht
https://soundcloud.com/begebot
https://soundcloud.com/web-standards
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw

Интересная статья о замыкании и лексическом окружении функций:
https://developer.mozilla.org/ru/docs/Web/JavaScript/Closures.

О пользовательских хуках:
https://ru.reactjs.org/docs/hooks-custom.html

https://reactdev.ru/
Справочник React

Какие бывают стили в названиях файлов js и зачем они нужны, интересно и кратко написано здесь:
https://medium.com/@alivander/camel-pascal-snake-case-и-другие-стили-написания-288ec62ca0d0

Ресурс по библиотеке Redux, которая очень широко используется в реальных проектах на React:
https://rajdee.gitbooks.io/redux-in-russian/content/

Создание приложений JavaScript с помощью TypeScript
https://docs.microsoft.com/ru-ru/learn/paths/build-javascript-applications-typescript/

Руководство Typescript
http://typescript-lang.ru/docs/

TypeScript в деталях.
https://typescript-definitive-guide.ru/book/chapters

**Желаю дальнейших творческих успехов, экспертных знаний и удачи!**

Pancfly commented:
Спасибо)
